jupyterhub_moss/models.py

```diff
@@ -0,0 +1,210 @@
+    """Information about partition description and available environments"""
+
+    architecture = ""
+    description = ""
+    jupyter_environments: Dict[str, JupyterEnvironment]
+    simple = True
+
+
+class PartitionInfo(
+    PartitionConfig, PartitionResources, allow_mutation=False, extra=Extra.allow
+):
+    """Complete information about a partition: config and resources"""
+
+    pass
+
+
+class _PartitionTraits(PartitionConfig, allow_mutation=False, extra=Extra.forbid):
+    """Configuration of a single partition passed as ``partitions`` traits"""
+
+    gpu: Optional[str] = None
+    max_ngpus: Optional[int] = None
+    max_nprocs: Optional[int] = None
+    max_runtime: Optional[int] = None
+
+    # validators
+    _is_positive = validator("max_ngpus", allow_reuse=True)(is_positive)
+    _check_match_gpu = validator("max_ngpus", allow_reuse=True)(check_match_gpu)
+    _is_strictly_positive = validator("max_nprocs", "max_runtime", allow_reuse=True)(
+        is_strictly_positive
+    )
+
+
+class PartitionsTrait(BaseModel, allow_mutation=False, extra=Extra.forbid):
+    """Configuration passed as ``partitions`` trait"""
+
+    __root__: Dict[str, _PartitionTraits]
+
+    def dict(self, *args, **kwargs):
+        return {k: v.dict(*args, **kwargs) for k, v in self.__root__.items()}
+
+    def items(self):
+        return self.__root__.items()
+
+
+class UserOptions(BaseModel):
+    """Options passed as `Spawner.user_options`"""
+
+    # Options received through the form or GET request
+    partition: str
+    runtime = ""
+    nprocs = 1
+    memory = ""
+    reservation = ""
+    ngpus = 0
+    options = ""
+    output = "/dev/null"
+    environment_path = ""
+    default_url = ""
+    root_dir = ""
+    # Extra fields
+    gres = ""
+    prologue = ""
+
+    @classmethod
+    def parse_formdata(cls, formdata: dict[str, list[str]]) -> UserOptions:
+        excluded_keys = "gres", "prologue"
+        fields = {
+            k: v[0].strip() for k, v in formdata.items() if k not in excluded_keys
+        }
+
+        # Compatibility with mem query param renamed memory
+        if "mem" in fields and "memory" not in fields:
+            fields["memory"] = fields["mem"]
+
+        # Convert output boolean query param to file pattern
+        fields["output"] = (
+            "slurm-%j.out" if fields.get("output", "false") == "true" else "/dev/null"
+        )
+        return cls.parse_obj(fields)
+
+    # validators
+    _is_positive = validator("ngpus", allow_reuse=True)(is_positive)
+    _is_strictly_positive = validator("nprocs", allow_reuse=True)(is_strictly_positive)
+
+    @validator(
+        "partition",
+        "runtime",
+        "memory",
+        "reservation",
+        "options",
+        "output",
+        "environment_path",
+        "default_url",
+        "root_dir",
+    )
+    def has_no_newline(cls, v: str) -> str:
+        if "\n" in v:
+            raise ValueError("Must not contain newline")
+        return v
```

Comment:
No, it's checking that there is not newline inside the field so one cannot use this to add lines to the sbatch script.
But it would also be good to strip leading/trailing whitspaces.